feat: support adding and fetching recurrence with `rrule`

differental opened the pull request:
Adds support of adding and fetching recurrence with [rrule](https://github.com/fmeringdal/rust-rrule), which follows `RFC-5545`. Tests are added as well.

Note: Presumably `src/period.rs` and `src/repeats.rs` can be removed if this is a satisfactory implementation.

differental committed:
feat: Add support for recurrence rules

differental committed:
test: add tests for recurrence

differental committed:
refactor: remove [0] indexing errors

qwandor commented on Cargo.toml:
```diff
@@ -25,6 +25,7 @@ serde = { version = "1.0", optional = true, features = ["derive"] }
 serde_json = { version = "1.0", optional = true }
 iso8601 = "0.6"
 chrono-tz = { version = "0.10", optional = true }
+rrule = "0.10.0"
```
0.13.0 is the latest version, any reason for using an old version here?

differental replied:
`v0.10.0` is the latest version supported by `rustc 1.60`, hence using the old version in this pr.

(Forcing `v0.13.0` did compile successfully on my machine though)

hoodie replied:
we don't have to stick wit that old MSRV. latest is 1.86, I think we can go with at least 1.80 by now

hoodie replied:
I think we can do it in here

qwandor commented on src/components.rs:
```diff
@@ -356,6 +357,27 @@ pub trait EventLike: Component {
     }
 
+    /// Set recurrence rules
+    fn recurrence(&mut self, rruleset: RRuleSet) -> &mut Self {
```
It looks like this ignores any `EXRULE`, `RDATE` or `EXDATE` from the set, which seems misleading. I think it should either add them as well (and maybe do something about the `DTSTART` matching?) or take an `RRule` rather than an `RRuleSet`.

differental replied:
`RDATE` and `EXDATE` added - thanks for spotting this. (`EXRULE` is deprecated)

For `DTSTART` matching, I can try adding it. The logic might become a bit complex though, since `rrule` takes a `chrono::DateTime<Tz>` and `icalendar` has a `DatePerhapsTime`. I would imagine it'll need some methods under `chrono-tz` flag, so I'm not sure whether it's best to be included in this pr as well... open to your thoughts. ;)

qwandor commented on src/components.rs:
```diff
@@ -356,6 +357,27 @@ pub trait EventLike: Component {
+    /// Get recurrence rules
+    fn get_recurrence(&self) -> Option<RRuleSet> {
+        let dt_start_str = self.property_value("DTSTART")?;
+        let rrule_str = self.property_value("RRULE")?;
```
What about `EXRULE`, `RDATE` and `EXDATE`?

differental replied:
Added `RDATE` and `EXDATE`, along with tests.

`EXRULE` seems to have been [deprecated](https://www.rfc-editor.org/rfc/rfc5545#appendix-A.3) as of rfc5545, so I didn’t include this. (It’s also under a feature flag in the `rrules` crate)

ProbstDJakob replied:
Shouldn't `EXRULE` still be passed (if the feature is enabled)?

differental commented:
Sorry for the oversight. `rrule v0.10.0` is the [latest](https://crates.io/crates/rrule/versions) supported by rustc 1.60, hence that was added when I ran `cargo add`.

v0.10.0 also did not put `RDATE`, `EXRULE` and `EXDATE` in `fmt`, leading me to believe everything's in `RRULE`.

I'll update my pr to include these three, hopefully v0.10.0 is sufficient.

hoodie commented:
Hi, first of all thanks for the contribution. I've been putting this off thinking we could maybe implement a native/own version of this here at some point, but now that I've looked at the rrule crate, there's a lot of stuff in there that I wouldn't want to debug myself 😆 

I welcome this addition, still, it pulls in a large dependency and change a few things. So let's be a bit careful. I would recommend making this an optional feature and making sure that it works together well with the parsing aspects of this library.

differental committed:
refactor: remove [0] indexing errors

differental committed:
refactor: remove [0] indexing errors

differental commented on src/components.rs:
```diff
@@ -1,4 +1,5 @@
 use chrono::{DateTime, NaiveDate, Utc};
+use rrule::RRuleSet;
```
I re-exported some common stuff (`RRuleSet`, `Frequency`, `NWeekday`, `RRule`, `Tz`, `Weekday`) in `lib.rs` under the feature flag - hope this is correct now. 

Thank you for your patience :)

hoodie replied:
I'm in no hurry here, take your time. Thanks for this contribution. I might also need a little more time to review the rest of this BTW

hoodie commented on src/lib.rs:
```diff
@@ -57,6 +57,9 @@ pub use crate::{
 };
 
+#[cfg(feature = "recurrence")]
+pub use rrule::{RRuleSet, Frequency, NWeekday, RRule, Tz, Weekday};
```
actually, maybe it is sufficient to reexport `rrule` itself. no need to make the api surface so large. accessing stuff like this is still convenient: `use icalendar::rrule::Weekday`.
That is how e.g. [axum](https://docs.rs/axum/latest/axum/) reexports [http](https://docs.rs/axum/latest/axum/#reexports) .

differental committed:
refactor: re-export `rrule` directly instead

differental committed:
refactor: Update MSRV to 1.80, update rrule to 0.13

differental committed:
refactor: update ci to include recurrence flag

differental committed:
refactor: remove debug msg and change deprecated methods

hoodie commented:
I was just playing with it *briefly*, I don't have a lot of time so I didn't get a good feel for the api yet. It might be that the API experience is still a bit complicated, when you have to give it a start data with the rrule tz. Also I wonder if an event itself that has a recurrence should maybe implement `ToIterator` or something?

Here is my short attempt at an example (5min efford).

```rs

use chrono::*;
use icalendar::*;

fn main() {
    let my_calendar = Calendar::new()
        .name("example calendar")
        .push(
            // add an all-day event
            Event::new()
                .all_day(NaiveDate::from_ymd_opt(2016, 3, 15).unwrap())
                .recurrence(
                    rrule::RRule::new(rrule::Frequency::Yearly)
                        .build(chrono::Utc::now().with_timezone(&rrule::Tz::UTC))
                        .unwrap(),
                )
                .summary("My Birthday")
                .description("Hey, I'm gonna have a party\nBYOB: Bring your own beer.\nHendrik")
                .done(),
        )
        .done();

    for event in my_calendar
        .components
        .first()
        .unwrap()
        .as_event()
        .unwrap()
        .get_recurrence()
        .iter()
    {
        println!("Event: {:?}", event);
    }
```

hoodie commented:
Would be nice if we could create a birthday event and then `.iter().take(5)` print the first 5 occurences of the event. or similar. I'm not saying you have to write this example, I'm just dumping my brain :D

differental committed:
test: add recurrence examples

differental commented:
The `recurrence_parse` example isn't working as intended - the example ics in `fixtures/icalendar-rb/recurrence.ics` seems to not have `EXDATE` immediately after `RRULE` and this is causing issues. I'll change the code to parse them separately

differental committed:
refactor: parse rdates and exdates separately

differental committed:
test: add extra tests for parsing recurrence

differental commented:
Oops sorry didn't see the comments - I added two examples that seem to work, and thankfully the parsing example with the recurrence fixture helped me discover the bug with parsing ics. Had to make some significant changes but now everything should work - I added an extra test just to be sure

differental committed:
refactor: remove redundant closures

hoodie commented:
No worries. Still grateful for the contribution. This is actually a complex feature. Let's take it slow.

differental commented:
It should work now though, I'll double-check the parsing just in case

differental committed:
test: add additional testcase for recurrence parsing

hoodie commented on examples/recurrence.rs:
```diff
@@ -0,0 +1,34 @@
+const RECURRENCE_LIMIT: u16 = u16::MAX;
+
+fn main() {
+    let dt_start = rrule::Tz::Europe__London
```
There are a few things that I'm not super happy with here.
1. having to pass a dt_start for the rrule and a configuring a separate start time with `.all_day()` probably confusing.
2. I just realized that rrule both depends on a different version of chrono-tz and even brings it's own `Tz` enum. That makes for a rather unfortunate api.

I'm wondering if we can maybe have `.recurrance()` accept an `rrule::Rule` instead of an `rrule::Ruleset` and construct that internally. That would require more internal logic, but I hope that the api improvement might be worth it.

Picture this instead:

```rs
    let my_event = Event::new()
        .all_day(Berlin.with_ymd_and_hms(2025, 3, 17, 0, 0, 0).unwrap())
        .summary("weekly event")
        .description("this event happens every Monday for four weeks")
        .recurrence(
            RRule::default()
                .count(4)
                .freq(rrule::Frequency::Weekly)
                .by_weekday(vec![rrule::NWeekday::Every(rrule::Weekday::Mon)]),
        ).unwrap()
        .done();
```

hoodie replied:
it is also important to hide the concrete datetime type if we want to support other datetime libs in the future. There have already been requests for [jiff](https://crates.io/crates/jiff)

hoodie replied:
alternatively we could also just have the `.recurrence()` method set the start time. It probably makes no sense to have _both_ `.starts()`/`.all_day()` and `.recurrence()` they could contradict each other. I need to think about this a little more.

Thanks for the examples so far. Helped me a lot 🙏

differental commented:
> How can I help with this? I would like to use recurrence in my own project and would love this PR to be merged.

Hi, sorry! I've been quite busy preparing for my exams for the past few weeks and I likely won't be available until 21st June. I'll change the API as soon as I'm available. 🥲

ProbstDJakob commented:
Of course, that's what I think as well. It could become a problem, but as long as it hasn't this should not be the main concern of this PR. But wrapping the `rrule` crate with own types and functions could not only benefit possible fixing the performance problem in the future and better ergonomics, but also decouple from breaking changes of the `rrule` crate.

ProbstDJakob commented on src/components.rs:
```diff
@@ -356,6 +358,69 @@ pub trait EventLike: Component {
+            .join(",");
+        if !exdates_str.is_empty() {
+            exdates_str = format!("\nEXDATE:{exdates_str}");
+        }
```
Currently the implementation does not pass parameters of the properties to the `rrule`. Either the valid parameters within the `rrule` should be included or if possible all parameters should be passed to the `format!` macro for example by using the `Property::fmt_write`/`<Property as TryInto<String>::try_into` and removing the trailing `\r\n` or by adding a `impl Display for Property` which does not add a `\r\n`.

The problem is, that timestamps which include a time zone but as `TZID` parameter get rendered as a "floating" date-time and thus use the time zone of the local machine which is incorrect. E.g. an event with `DTSTART;TZID=Europe/Berlin:20251006T083000` will be rendered as `DTSTART:20251006T083000
RRULE:[...]` by this function.

ProbstDJakob replied:
For example this works: https://github.com/ProbstDJakob/icalendar/commit/58d99eece50fc72070419e92744e359afd84841f

ProbstDJakob replied:
For example this works: https://github.com/ProbstDJakob/icalendar/commit/58d99eece50fc72070419e92744e359afd84841f

ProbstDJakob replied:
The implementation still does not pass parameters and thus the `TZID` and `VALUE`.

ProbstDJakob commented on src/components.rs:
```diff
@@ -356,6 +358,69 @@ pub trait EventLike: Component {
+        }
+
+        let rrules = format!("DTSTART:{dt_start_str}\nRRULE:{rrule_str}{rdates_str}{exdates_str}");
+        rrules.parse::<RRuleSet>().ok()
```
It may be better to return the error (or wrap it), so that the caller can check whether there was an error parsing the `RRULE`/there is no `RRULE` or `DTSTART` was missing.

hoodie committed:
refactor: change get_recurrence to return Option

hoodie committed:
refactor: use unbuilt RRule

instead of a built `RRuleSet`. Callers must build the rule inside the
method.

hoodie committed:
refactor: reexport selected rrule types

hoodie committed:
build: add chrono-tz feature and enable in example

hoodie committed:
build: fix example

hoodie committed:
refactor: simplify get_recurrence API

hoodie committed:
refactor: add type alias for Unvalidated RRule

hoodie committed:
refactor: add a custom recurrenace error wrapper

ProbstDJakob commented:
I was a bit busy during the week (and still am until next week), but if you are still interested in feedback (and so do I read your message), I would try to take some time to review it. From what I read so far, you still expose the `rrule` crate and drop most/all parameters of `DTSTART`, `RDATE`, and `EXDATE` when passing them to the `rrule` crate. I am not that familiar with the RFC (and would take a closer look, when reviewing your PR), but I do not think it is wise to just remove them.

ProbstDJakob commented on examples/recurrence_parse.rs:
```diff
@@ -0,0 +1,25 @@
+        if let CalendarComponent::Event(event) = component {
+            println!("Event: {}", event.get_summary().unwrap());
+            if let Some(rrules) = event.get_recurrence() {
+                let datetimes: Vec<DateTime<Tz>> = rrules.all(RECURRENCE_LIMIT).dates;
```
Loading `u16::MAX` elements into a vector is (generally) not a good idea if only a subset is needed. So something like `rrule.into_iter().take(10).collect()` could be used.

ProbstDJakob commented on src/components.rs:
```diff
@@ -437,6 +441,140 @@ pub trait EventLike: Component {
+            }
+            Some(DatePerhapsTime::DateTime(CalendarDateTime::Floating(naive))) => {
+                // Floating datetimes have no timezone — use UTC as a best-effort interpretation.
+                rrule::Tz::UTC.from_utc_datetime(&naive)
```
The date-time does not have a time-zone info and thus should be handled as such. Therefore `rrule::Tz::LOCAL` should probably be used.

ProbstDJakob commented on src/components.rs:
```diff
@@ -437,6 +441,140 @@ pub trait EventLike: Component {
+            }
+            Some(DatePerhapsTime::Date(naive_date)) => {
+                // All-day events: treat as midnight UTC.
+                rrule::Tz::UTC.from_utc_datetime(&naive_date.and_hms_opt(0, 0, 0).unwrap())
```
I am not sure if the RFC5545 also allows dates with time-zone info (could neither find a sentence allowing it neither denying, only the format definition which at least does not exclude the possibility), but at least this case should probably also be handled with `rrule::Tz::LOCAL`. Best case would be if `rrule` would support a simple date for `RRule::build` so it would also report only dates when collection the recurrences.

ProbstDJakob commented on src/components.rs:
```diff
@@ -437,6 +441,140 @@ pub trait EventLike: Component {
+
+        let rruleset = rrule.build(dt_start)?;
+
+        let rrule_str = rruleset
+            .get_rrule()
+            .iter()
+            .map(ToString::to_string)
+            .collect::<Vec<_>>()
+            .join("\n");
+
+        let rdates = rruleset
+            .get_rdate()
+            .iter()
+            .map(|dt| dt.format("%Y%m%dT%H%M%SZ").to_string())
+            .collect::<Vec<_>>()
+            .join(",");
+
+        let exdates = rruleset
+            .get_exdate()
+            .iter()
+            .map(|dt| dt.format("%Y%m%dT%H%M%SZ").to_string())
+            .collect::<Vec<_>>()
+            .join(",");
+
+        self.add_property("RRULE", rrule_str)
+            .add_multi_property("RDATE", &rdates)
+            .add_multi_property("EXDATE", &exdates);
```
In comparison to what `<RRuleSet as Display>::display` does, this lacks some cases. Especially the complete `EXRULE` part, but also some time-zone related stuff. I've read your comment that `EXRULE` is deprecated, so maybe just check if those are set and if so return an error (and/or state it within the documentation).

ProbstDJakob reviewed (commented):
Much work and no time, but I finally got to look into it. I've neither looked through the tests nor the PR #169 but the later I will review next.
